hexrd/sampleOrientations/sampleRFZ.py

```diff
@@ -0,0 +1,162 @@
+    returns a list of orientations which are spaced,
+    on an average, to the user specified angular spacing
+    @author Saransh Singh, LLNL
+    @date   11/28/2022 1.0 original
+
+     Note
+    ----
+    Details can be found in:
+    S. Singh and M. De Graef, "Orientation sampling for 
+    dictionary-based diffraction pattern indexing methods". 
+    MSMSE 24, 085013 (2016)
+
+    Attributes
+    ----------
+    pgnum: int
+        point group number of crystal
+    average_angular_spacing" float
+        average angular spacing of sampling (in degrees)
+
+    """
+
+    def __init__(self,
+                 pgnum,
+                 sampling_type='default',
+                 average_angular_spacing=3.0):
+        """__init__ method of the sampleRFZ class.
+
+
+        Parameters
+        ----------
+        pgnum : int
+            point group number
+        sampling_type : str
+            default sampling with origin
+            special which is shifted to mid-points
+            of each cubochoric cell
+        average_angular_spacing : float
+            average angular spacing in degrees
+
+        """
+
+        self.ap_2 = constants.cuA_2
+        self.pgnum = pgnum
+        self.sampling_type = sampling_type
+        self.avg_ang_spacing = average_angular_spacing
+
+    def sampling_N(self):
+        """Get the number of sampling steps in the cubochoric
+        cube based on the average angular spacing requested.
+        Uses eqns. 9 and 10 of S. Singh and M. De Graef MSMSE 24,
+        085013 (2016)
+
+        """
+        if self.sampling_type.lower() == 'default':
+            return np.rint(131.97049 / (self.avg_ang_spacing - 0.03732)).astype(np.int32)
+        elif self.sampling_type.lower() == 'special':
+            return np.rint(125.70471 / (self.avg_ang_spacing - 0.07127)).astype(np.int32)
+
+    def sample(self):
+        self.orientations = _sample(self.pgnum,
+                                    self.cubN,
+                                    self.delta,
+                                    self.shift,
+                                    self.ap_2)
+
+    @property
+    def pgnum(self):
+        return self._pgnum
+
+    @pgnum.setter
+    def pgnum(self, pgn):
+        self._pgnum = pgn
+        if hasattr(self, 'sampling_type'):
+            if hasattr(self, 'avg_ang_spacing'):
+                self.sample()
+
+
+    @property
+    def sampling_type(self):
+        return self._sampling_type
+
+    @sampling_type.setter
+    def sampling_type(self, stype):
+        self._sampling_type = stype
+        if hasattr(self, 'pgnum'):
+            if hasattr(self, 'avg_ang_spacing'):
+                self.sample()
+
+    @property
+    def avg_ang_spacing(self):
+        return self._avg_ang_spacing
+
+
+    @avg_ang_spacing.setter
+    def avg_ang_spacing(self, ang):
+        self._avg_ang_spacing = ang
+        if hasattr(self, 'pgnum'):
+            if hasattr(self, 'sampling_type'):
+                self.sample()
```

Comment:
One alternative you could have for all of these `if hasattr()` checks is to make a separate function like this:

```python
    def sample_if_possible(self):
        required_attributes = ('pgnum', 'avg_ang_spacing', 'sampling_type')
        if not all(hasattr(self, x) for x in required_attributes):
            return

        self.sample()
```

And then in all of the setters after setting, just call this function. For instance:

```python
    @avg_ang_spacing.setter
    def avg_ang_spacing(self, ang):
        self._avg_ang_spacing = ang
        self.sample_if_possible()
```

Reply:
Yes, the code looks much cleaner with this. I'll apply these changes.